tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc

```diff
@@ -0,0 +1,159 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <utility>
+
+#include "tensorflow/compiler/xla/service/gpu/tests/gpu_codegen_test.h"
+#include "tensorflow/compiler/xla/service/hlo_module_config.h"
+#include "tensorflow/compiler/xla/service/hlo_parser.h"
+#include "tensorflow/compiler/xla/tests/hlo_test_base.h"
+#include "tensorflow/core/platform/test.h"
+
+namespace xla {
+namespace gpu {
+namespace {
+
+class GpuSliceInputFusionTest : public GpuCodegenTest {
+ protected:
+  GpuSliceInputFusionTest() {}
+
+  // Disable the layout assignment pass as it would throw away the layouts
+  // in the fusion computation.
+  HloModuleConfig ConfigWithoutLayoutAssignment() {
+    HloModuleConfig config;
+    auto debug_options = HloTestBase::GetDebugOptionsForTest();
+    // Disable layout_assignment to use the preassigned layouts.
+    debug_options.add_xla_disable_hlo_passes("layout-assignment");
+    config.set_debug_options(debug_options);
+    return config;
+  }
+};
+
+TEST_F(GpuSliceInputFusionTest, InputFusionWithOnlyOneSlice) {
+  const char *const kHloString = R"(
+  HloModule input_fusion_with_only_one_slice
+
+  fused_computation {
+    arg.1 = f16[1024,512]{1,0} parameter(0)
+    arg.2 = f16[1024,512]{1,0} parameter(1)
+    arg1.conv = f32[1024,512]{1,0} convert(arg.1)
+    arg2.conv = f32[1024,512]{1,0} convert(arg.2)
+    add.1 = f32[1024,512]{1,0} add(arg1.conv, arg2.conv)
+    ROOT slice.1 = f32[512,511]{1,0} slice(add.1), slice={[512:1024], [1:512]}
+  }
+
+  ENTRY kernel_entry {
+    arg.1 = f16[1024,512]{1,0} parameter(0)
+    arg.2 = f16[1024,512]{1,0} parameter(1)
+    ROOT fusion = f32[512, 511]{1,0} fusion(arg.1, arg.2), kind=kInput,
+        calls=fused_computation
+  })";
+
+  auto hlo_module =
+      ParseAndReturnVerifiedModule(kHloString, ConfigWithoutLayoutAssignment())
+          .ValueOrDie();
+  CompileAndVerifyIr(std::move(hlo_module),
+                     R"(
+; CHECK-LABEL: define void @fusion
```

Comment:
Could we have a slightly more involved codegen test? This commit adds quite a lot of codegen logic.

Reply:
Do you mean to add a more involved test pattern or you feel the IR CHECK is not enough?

Reply:
Yes, a more involved pattern would be nice.

Reply:
I gave some more contexts in the main discussion thread. I will wait for your response there.